### PR TITLE
Venue and Organizer archives base support

### PR DESCRIPTION
**Pull request by lucatume**
Work in progress.

Ticket: https://central.tri.be/issues/81510

Builds on: https://github.com/moderntribe/tribe-common/pull/431

This PR adds the code needed to support Venue and Organizer archives:
* add Venue and Organizer archives
* support basic archive functions and filtering:
	* by search (on all post and custom fields)
	* by upcoming events
	* by `has_no_events` criteria
	* by event
* extended the functionality of the `Tribe__Venue` class to improve custom field queries and make it suitable for search

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:399](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L399)
:warning: | Comment refers to a TODO task &quot;Implement get_documentation() method&quot; | [Archive_Venue.php:36](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L36)
:x: | There must be a single space after the first parenthesis | [Archive_Venue.php:159](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L159)
:x: | There must be a single space before the last parenthesis | [Archive_Venue.php:159](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L159)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:165](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L165)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:197](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L197)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:223](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L223)
:warning: | The method parameter $only_with_upcoming is never used | [venue.php:390](https://github.com/moderntribe/the-events-calendar/blob/4742202bf213cd8bbab2aebe3b1b520a4893f64b/src/functions/template-tags/venue.php#L390)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:399](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L399)
:warning: | Comment refers to a TODO task &quot;Implement get_documentation() method&quot; | [Archive_Venue.php:31](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L31)
:x: | There must be a single space after the first parenthesis | [Archive_Venue.php:188](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L188)
:x: | There must be a single space before the last parenthesis | [Archive_Venue.php:188](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L188)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:194](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L194)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:226](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L226)
:x: | Each line in an array declaration must end in a comma | [Archive_Venue.php:252](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L252)
:x: | Comma required after last value in array declaration | [Venue.php:126](https://github.com/moderntribe/the-events-calendar/blob/952ac27547ed05334995d9cd25f180051ec1f85e/src/Tribe/Venue.php#L126)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Comment refers to a TODO task &quot;Implement get_documentation() method&quot; | [Archive_Venue.php:31](https://github.com/moderntribe/the-events-calendar/blob/20dcfae34481f208daefd4ce592c9ffc9fec1e91/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L31)
:x: | Comma not allowed after last value in single-line array declaration | [Venue.php:124](https://github.com/moderntribe/the-events-calendar/blob/20dcfae34481f208daefd4ce592c9ffc9fec1e91/src/Tribe/Venue.php#L124)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | There must be a single space after the first square bracket | [Base.php:102](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L102)
:x: | There must be a single space before the last square bracket | [Base.php:102](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L102)
:x: | Expected 1 space before &quot;=&quot;; 0 found | [Base.php:102](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L102)
:x: | Expected 1 space after &quot;=&quot;; 0 found | [Base.php:102](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L102)
:x: | There must be a single space after the first parenthesis | [Base.php:173](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L173)
:x: | There must be a single space before the last parenthesis | [Base.php:173](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L173)
:x: | There must be a single space before the last parenthesis | [Base.php:173](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L173)
:x: | Missing space before array closer. | [Base.php:173](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Linked_Posts/Base.php#L173)
:x: | There must be a space between if and the first parenthesis '(' | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | No space after opening parenthesis is prohibited | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | ! must be followed by a single space | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | No space before closing parenthesis is prohibited | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | You must have a space after a closing parenthesis | [Archive_Organizer.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Organizer.php#L94)
:x: | There must be a space between if and the first parenthesis '(' | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | No space after opening parenthesis is prohibited | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | ! must be followed by a single space | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space after the first parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | There must be a single space before the last parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | No space before closing parenthesis is prohibited | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | You must have a space after a closing parenthesis | [Archive_Venue.php:94](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/REST/V1/Endpoints/Archive_Venue.php#L94)
:x: | Comma not allowed after last value in single-line array declaration | [Venue.php:124](https://github.com/moderntribe/the-events-calendar/blob/a45db433d22479f9d15e4f1726bddb9d064fa57f/src/Tribe/Venue.php#L124)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Linked_Posts/Base.php [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+		}
+
+		$found               = array_map( 'intval', array_values( $found ) );
+
+		$cache[ $cache_key ] = $found;
+
+		return $found;
+	}
+
+	public function find_with_upcoming_events( $only_with_upcoming = true ) {
```
Comment (borkweb) [inline src/Tribe/Linked_Posts/Base.php]: Add docblock?

File: src/Tribe/Organizer.php [hunk 2]
```diff
@@ -547,27 +564,10 @@ protected function get_duplicate_custom_fields() {
 		 *
 		 * @param array $fields An array associating the custom field meta key to the strategy definition.
 		 *
-		 * @see   Tribe__Duplicate__Strategy_Factory
+		 * @see   Tribe__Duplicrte__Strategy_Factory
```
Comment (borkweb) [inline src/Tribe/Organizer.php]: `_Duplicate_` vs `_Duplicrte_`

File: src/Tribe/REST/V1/Endpoints/Archive_Event.php [hunk 3]
```diff
@@ -508,22 +387,29 @@ public function READ_args() {
...
 	 */
-	protected function parse_args( array $args, array $defaults ) {
-		foreach ( $this->supported_query_vars as $request_key => $query_var ) {
-			if ( isset( $defaults[ $request_key ] ) ) {
-				$defaults[ $query_var ] = $defaults[ $request_key ];
-			}
-		}
+	protected function get_total( $args ) {
+		$this->total = count( tribe_get_events( array_merge( $args, array(
+			'posts_per_page'         => - 1,
```
Comment (borkweb) [inline src/Tribe/REST/V1/Endpoints/Archive_Event.php]: Not necessary to change, but the space between the `-` and the `1` is kind of weird.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | TRUE, FALSE and NULL must be lowercase; expected &quot;false&quot; but found &quot;FALSE&quot; | [Archive_Event.php:398](https://github.com/moderntribe/the-events-calendar/blob/e0f8d067804bade86c9851c51101e9a3c183efce/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L398)
:x: | TRUE, FALSE and NULL must be lowercase; expected &quot;false&quot; but found &quot;FALSE&quot; | [Archive_Event.php:399](https://github.com/moderntribe/the-events-calendar/blob/e0f8d067804bade86c9851c51101e9a3c183efce/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L399)

> via jenkins/codesniffer-phpcs

**Pull request merged**